rust-symcrypt/src/ecc/mod.rs

```diff
@@ -0,0 +1,587 @@
+pub enum EcKeyUsage {
+    /// When using [`EcKeyUsage::EcDsa`] the intended usage of the key will be only for EcDsa sign or verify.
+    EcDsa,
+
+    /// When using [`EcKeyUsage::EcDh`] the intended usage of the key will be only for EcDh.
+    EcDh,
+
+    /// When using [`EcKeyUsage::EcDhAndEcDsa`] the intended usage of the key will be for both EcDsa and EcDh.
+    EcDhAndEcDsa,
+}
+
+// to_symcrypt_flag converts the EcKeyUsage enum to the corresponding SymCrypt flag and is only needed internally.
+impl EcKeyUsage {
+    pub(crate) fn to_symcrypt_flag(&self) -> u32 {
+        match self {
+            EcKeyUsage::EcDsa => symcrypt_sys::SYMCRYPT_FLAG_ECKEY_ECDSA,
+            EcKeyUsage::EcDh => symcrypt_sys::SYMCRYPT_FLAG_ECKEY_ECDH,
+            EcKeyUsage::EcDhAndEcDsa =>  { symcrypt_sys::SYMCRYPT_FLAG_ECKEY_ECDSA | symcrypt_sys::SYMCRYPT_FLAG_ECKEY_ECDH },
+        }
+    }
+}
+
+#[derive(Debug)]
+/// ECC Key State Object.
+/// 
+/// `EcKey` is a struct that represents an Ec Key. It can hold either just an Ec public key, or an Ec key pair.
+/// To check if if the key has a private key attached, you can use [`EcKey::has_private_key()`]
+pub struct EcKey {
+    inner_key: InnerEcKey,
+    curve_type: CurveType,
+    ec_key_usage: EcKeyUsage,
+    has_private_key: bool
+}
+
+unsafe impl Send for EcKey {
+    // TODO: Discuss send/sync for rustls
+}
+
+unsafe impl Sync for EcKey {
+    // TODO: Discuss send/sync for rustls
+}
+
+// Must drop the EcKey before the expanded EcCurve is dropped
+// EcCurve has static lifetime so this will always be the case.
+impl Drop for EcKey {
+    fn drop(&mut self) {
+        unsafe {
+            // SAFETY: FFI calls
+            symcrypt_sys::SymCryptEckeyFree(self.inner_key());
+        }
+    }
+}
+
+impl EcKey {
+    /// `generate_key_pair()` generates a random Ec Key object that has a public and private key attached based on the [`CurveType`] and [`EcKeyUsage`] provided.
+    /// 
+    /// `curve_type` represents the [`CurveType`] of the key.
+    /// 
+    /// `ec_key_usage` represents the [`EcKeyUsage`] of the key and indicates if the key will be used for [`EcKeyUsage::EcDsa`] or [`EcKeyUsage::EcDh`], or [`EcKeyUsage::EcDhAndEcDsa`].
+    /// 
+    /// This function will return a [`SymCryptError::MemoryAllocationFailure`] if there is not enough memory to allocate the key.
+    pub fn generate_key_pair(curve_type: CurveType, ec_key_usage: EcKeyUsage) -> Result<Self, SymCryptError> {
+        let ec_curve = InnerEcCurve::new(curve_type); // Can fail here due to insufficient memory.
+        unsafe {
+            // SAFETY: FFI calls
+            // Stack allocated since we will do SymCryptEckeyAllocate.
+            let key_ptr = symcrypt_sys::SymCryptEckeyAllocate(ec_curve.0);
+            if key_ptr.is_null() {
+                return Err(SymCryptError::MemoryAllocationFailure);
+            }
```

Comment:
Need wrapper for this to avoid leaking SymCrypt allocated memory. Applies to all calls of SymCryptEckeyAllocat